File: application/src/main/java/com/github/cereda/arara/utils/DirectiveUtils.java
/**
 * Arara, the cool TeX automation tool
 * Copyright (c) 2012, Paulo Roberto Massa Cereda 
 * All rights reserved.
 *
 * Redistribution and  use in source  and binary forms, with  or without
 * modification, are  permitted provided  that the  following conditions
 * are met:
 *
 * 1. Redistributions  of source  code must  retain the  above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form  must reproduce the above copyright
 * notice, this list  of conditions and the following  disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither  the name  of the  project's author nor  the names  of its
 * contributors may be used to  endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS  PROVIDED BY THE COPYRIGHT  HOLDERS AND CONTRIBUTORS
 * "AS IS"  AND ANY  EXPRESS OR IMPLIED  WARRANTIES, INCLUDING,  BUT NOT
 * LIMITED  TO, THE  IMPLIED WARRANTIES  OF MERCHANTABILITY  AND FITNESS
 * FOR  A PARTICULAR  PURPOSE  ARE  DISCLAIMED. IN  NO  EVENT SHALL  THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY,  OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT  NOT LIMITED  TO, PROCUREMENT  OF SUBSTITUTE  GOODS OR  SERVICES;
 * LOSS  OF USE,  DATA, OR  PROFITS; OR  BUSINESS INTERRUPTION)  HOWEVER
 * CAUSED AND  ON ANY THEORY  OF LIABILITY, WHETHER IN  CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY  OUT  OF  THE USE  OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.github.cereda.arara.utils;

import com.github.cereda.arara.controller.ConfigurationController;
import com.github.cereda.arara.controller.LanguageController;
import com.github.cereda.arara.model.AraraException;
import com.github.cereda.arara.model.Conditional;
import com.github.cereda.arara.model.Directive;
import com.github.cereda.arara.model.Messages;
import com.github.cereda.arara.model.Pair;
import java.io.File;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.yaml.snakeyaml.DumperOptions;
import org.yaml.snakeyaml.Yaml;
import org.yaml.snakeyaml.constructor.Constructor;
import org.yaml.snakeyaml.error.MarkedYAMLException;
import org.yaml.snakeyaml.representer.Representer;

/**
 * Implements directive utilitary methods.
 * @author Paulo Roberto Massa Cereda
 * @version 4.0
 * @since 4.0
 */
public class DirectiveUtils {

    // the application messages obtained from the
    // language controller
    private static final LanguageController messages =
            LanguageController.getInstance();
    
    // get the logger context from a factory
    private static final Logger logger =
            LoggerFactory.getLogger(DirectiveUtils.class);

    /**
     * Extracts a list of directives from a list of strings.
     * @param lines List of strings.
     * @return A list of directives.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */    
    public static List<Directive> extractDirectives(List<String> lines)
            throws AraraException {

        boolean header = (Boolean) ConfigurationController.
                getInstance().get("execution.header");
        String regex = (String) ConfigurationController.
                getInstance().get("execution.file.pattern");
        Pattern linecheck = Pattern.compile(regex);
        regex = regex.concat((String) ConfigurationController.
                getInstance().get("application.pattern"));
        Pattern pattern = Pattern.compile(regex);
        List<Pair<Integer, String>> pairs =
                new ArrayList<Pair<Integer, String>>();
        Matcher matcher;
        for (int i = 0; i < lines.size(); i++) {
            matcher = pattern.matcher(lines.get(i));
            if (matcher.find()) {
                String line = lines.get(i).substring(
                        matcher.end(),
                        lines.get(i).length()
                );
                Pair<Integer, String> pair =
                        new Pair<Integer, String>(i + 1, line);
                pairs.add(pair);

                logger.info(
                        messages.getMessage(
                                Messages.LOG_INFO_POTENTIAL_PATTERN_FOUND,
                                (i + 1),
                                line.trim()
                        )
                );
            }
            else {
                if (header) {
                    if (!checkLinePattern(linecheck, lines.get(i))) {
                        break;
                    }
                }
            }
        }

        if (pairs.isEmpty()) {
            throw new AraraException(
                    messages.getMessage(
                            Messages.ERROR_VALIDATE_NO_DIRECTIVES_FOUND
                    )
            );
        }

        List<DirectiveAssembler> assemblers
                = new ArrayList<DirectiveAssembler>();
        DirectiveAssembler assembler = new DirectiveAssembler();
        regex = (String) ConfigurationController.getInstance().
                get("directives.linebreak.pattern");
        pattern = Pattern.compile(regex);
        for (Pair<Integer, String> pair : pairs) {
            matcher = pattern.matcher(pair.getSecondElement());
            if (matcher.find()) {
                if (!assembler.isAppendAllowed()) {
                    throw new AraraException(
                            messages.getMessage(
                                    Messages.ERROR_VALIDATE_ORPHAN_LINEBREAK,
                                    pair.getFirstElement()
                            )
                    );
                } else {
                    assembler.addLineNumber(pair.getFirstElement());
                    assembler.appendLine(matcher.group(1));
                }
            } else {
                if (assembler.isAppendAllowed()) {
                    assemblers.add(assembler);
                }
                assembler = new DirectiveAssembler();
                assembler.addLineNumber(pair.getFirstElement());
                assembler.appendLine(pair.getSecondElement());
            }
        }
        if (assembler.isAppendAllowed()) {
            assemblers.add(assembler);
        }

        List<Directive> directives = new ArrayList<Directive>();
        for (DirectiveAssembler current : assemblers) {
            directives.add(generateDirective(current));
        }
        return directives;

    }

    /**
     * Generates a directive from a directive assembler.
     * @param assembler The directive assembler.
     * @return The corresponding directive.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    public static Directive generateDirective(DirectiveAssembler assembler)
            throws AraraException {
        String regex = (String) ConfigurationController.getInstance().
                get("directives.pattern");
        Pattern pattern = Pattern.compile(regex);
        Matcher matcher = pattern.matcher(assembler.getText());
        if (matcher.find()) {
            Directive directive = new Directive();
            directive.setIdentifier(matcher.group(1));
            directive.setParameters(
                    getParameters(matcher.group(3), assembler.getLineNumbers())
            );
            Conditional conditional = new Conditional();
            conditional.setType(getType(matcher.group(5)));
            conditional.setCondition(getCondition(matcher.group(6)));
            directive.setConditional(conditional);
            directive.setLineNumbers(assembler.getLineNumbers());

            logger.info(
                    messages.getMessage(
                            Messages.LOG_INFO_POTENTIAL_DIRECTIVE_FOUND,
                            directive
                    )
            );

            return directive;
        } else {
            throw new AraraException(
                    messages.getMessage(
                            Messages.ERROR_VALIDATE_INVALID_DIRECTIVE_FORMAT,
                            CommonUtils.getCollectionElements(
                                    assembler.getLineNumbers(),
                                    "(",
                                    ")",
                                    ", "
                            )
                    )
            );
        }
    }

    /**
     * Gets the conditional type based on the input string.
     * @param text The input string.
     * @return The conditional type.
     */
    private static Conditional.ConditionalType getType(String text) {
        if (text == null) {
            return Conditional.ConditionalType.NONE;
        } else {
            if (text.equals("if")) {
                return Conditional.ConditionalType.IF;
            } else {
                if (text.equals("while")) {
                    return Conditional.ConditionalType.WHILE;
                } else {
                    if (text.equals("until")) {
                        return Conditional.ConditionalType.UNTIL;
                    } else {
                        return Conditional.ConditionalType.UNLESS;
                    }
                }
            }
        }
    }

    /**
     * Gets the condition from the input string.
     * @param text The input string.
     * @return A string representing the condition.
     */
    private static String getCondition(String text) {
        return text == null ? "" : text;
    }

    /**
     * Gets the parameters from the input string.
     * @param text The input string.
     * @param numbers The list of line numbers.
     * @return A map containing the directive parameters.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    private static Map<String, Object> getParameters(String text,
            List<Integer> numbers) throws AraraException {

        if (text == null) {
            return new HashMap<String, Object>();
        }

        Yaml yaml = new Yaml(
                new Constructor(),
                new Representer(),
                new DumperOptions(),
                new DirectiveResolver()
        );
        try {
            @SuppressWarnings("unchecked")
            HashMap<String, Object> map = yaml.loadAs(text, HashMap.class);
            return map;
        } catch (MarkedYAMLException exception) {
            throw new AraraException(
                    messages.getMessage(
                            Messages.ERROR_VALIDATE_YAML_EXCEPTION,
                            CommonUtils.getCollectionElements(
                                    numbers,
                                    "(",
                                    ")",
                                    ", "
                            )
                    ),
                    exception
            );
        }
    }

    /**
     * Validates the list of directives, returning a new list.
     * @param directives The list of directives.
     * @return A new list of directives.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    public static List<Directive> validate(List<Directive> directives)
            throws AraraException {

        ArrayList<Directive> result = new ArrayList<Directive>();
        for (Directive directive : directives) {
            Map<String, Object> parameters = directive.getParameters();

            if (parameters.containsKey("file")) {
                throw new AraraException(
                        messages.getMessage(
                                Messages.ERROR_VALIDATE_FILE_IS_RESERVED,
                                CommonUtils.getCollectionElements(
                                        directive.getLineNumbers(),
                                        "(",
                                        ")",
                                        ", "
                                )
                        )
                );
            }

            if (parameters.containsKey("reference")) {
                throw new AraraException(
                        messages.getMessage(
                                Messages.ERROR_VALIDATE_REFERENCE_IS_RESERVED,
                                CommonUtils.getCollectionElements(
                                        directive.getLineNumbers(),
                                        "(",
                                        ")",
                                        ", "
                                )
                        )
                );
            }

            if (parameters.containsKey("files")) {

                Object holder = parameters.get("files");
                if (holder instanceof List) {
                    @SuppressWarnings("unchecked")
                    List<Object> files = (List<Object>) holder;
                    parameters.remove("files");
                    if (files.isEmpty()) {
                        throw new AraraException(
                                messages.getMessage(
                                        Messages.ERROR_VALIDATE_EMPTY_FILES_LIST,
                                        CommonUtils.getCollectionElements(
                                                directive.getLineNumbers(),
                                                "(",
                                                ")",
                                                ", "
                                        )
                                )
                        );
                    }
                    for (Object file : files) {
                        Map<String, Object> map = new HashMap<String, Object>();
                        for (String key : parameters.keySet()) {
                            map.put(key, parameters.get(key));
                        }
                        File representation = CommonUtils.
                                getCanonicalFile(String.valueOf(file));

                        map.put("reference", representation);
                        map.put("file", representation.getName());

                        Directive addition = new Directive();
                        Conditional conditional = new Conditional();
                        conditional.setCondition(directive.getConditional().
                                getCondition()
                        );
                        conditional.setType(directive.getConditional().
                                getType()
                        );
                        addition.setIdentifier(directive.getIdentifier());
                        addition.setConditional(conditional);
                        addition.setParameters(map);
                        addition.setLineNumbers(directive.getLineNumbers());
                        result.add(addition);
                    }
                } else {
                    throw new AraraException(
                            messages.getMessage(
                                    Messages.ERROR_VALIDATE_FILES_IS_NOT_A_LIST,
                                    CommonUtils.getCollectionElements(
                                            directive.getLineNumbers(),
                                            "(",
                                            ")",
                                            ", "
                                    )
                            )
                    );
                }
            } else {
                File representation = (File) ConfigurationController.
                        getInstance().get("execution.reference");
                parameters.put("file", representation.getName());
                parameters.put("reference", representation);
                directive.setParameters(parameters);
                result.add(directive);
            }
        }

        logger.info(
                messages.getMessage(
                        Messages.LOG_INFO_VALIDATED_DIRECTIVES
                )
        );
        logger.info(DisplayUtils.displayOutputSeparator(
                messages.getMessage(
                        Messages.LOG_INFO_DIRECTIVES_BLOCK
                )
        ));
        for (Directive directive : result) {
            logger.info(directive.toString());
        }

        logger.info(DisplayUtils.displaySeparator());

        return result;
    }
    
    /**
     * Checks if the provided line contains the corresponding pattern, based on
     * the file type, or an empty line.
     * @param pattern Pattern to be matched, based on the file type.
     * @param line Provided line.
     * @return Logical value indicating if the provided line contains the
     * corresponding pattern, based on the file type, or an empty line.
     */
    private static boolean checkLinePattern(Pattern pattern, String line) {
        return CommonUtils.checkEmptyString(line.trim()) ||
                pattern.matcher(line).find();
    }

}

File: application/src/main/kotlin/com/github/cereda/arara/ruleset/DirectiveUtils.kt
/*
 * Arara, the cool TeX automation tool
 * Copyright (c) 2012 -- 2019, Paulo Roberto Massa Cereda
 * All rights reserved.
 *
 * Redistribution and  use in source  and binary forms, with  or without
 * modification, are  permitted provided  that the  following conditions
 * are met:
 *
 * 1. Redistributions  of source  code must  retain the  above copyright
 * notice, this list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form  must reproduce the above copyright
 * notice, this list  of conditions and the following  disclaimer in the
 * documentation and/or other materials provided with the distribution.
 *
 * 3. Neither  the name  of the  project's author nor  the names  of its
 * contributors may be used to  endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS  PROVIDED BY THE COPYRIGHT  HOLDERS AND CONTRIBUTORS
 * "AS IS"  AND ANY  EXPRESS OR IMPLIED  WARRANTIES, INCLUDING,  BUT NOT
 * LIMITED  TO, THE  IMPLIED WARRANTIES  OF MERCHANTABILITY  AND FITNESS
 * FOR  A PARTICULAR  PURPOSE  ARE  DISCLAIMED. IN  NO  EVENT SHALL  THE
 * COPYRIGHT HOLDER OR CONTRIBUTORS BE  LIABLE FOR ANY DIRECT, INDIRECT,
 * INCIDENTAL, SPECIAL, EXEMPLARY,  OR CONSEQUENTIAL DAMAGES (INCLUDING,
 * BUT  NOT LIMITED  TO, PROCUREMENT  OF SUBSTITUTE  GOODS OR  SERVICES;
 * LOSS  OF USE,  DATA, OR  PROFITS; OR  BUSINESS INTERRUPTION)  HOWEVER
 * CAUSED AND  ON ANY THEORY  OF LIABILITY, WHETHER IN  CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY
 * WAY  OUT  OF  THE USE  OF  THIS  SOFTWARE,  EVEN  IF ADVISED  OF  THE
 * POSSIBILITY OF SUCH DAMAGE.
 */
package com.github.cereda.arara.ruleset

import com.github.cereda.arara.Arara
import com.github.cereda.arara.configuration.AraraSpec
import com.github.cereda.arara.localization.LanguageController
import com.github.cereda.arara.localization.Messages
import com.github.cereda.arara.model.AraraException
import com.github.cereda.arara.utils.CommonUtils
import com.github.cereda.arara.utils.DisplayUtils
import org.slf4j.LoggerFactory
import org.yaml.snakeyaml.DumperOptions
import org.yaml.snakeyaml.Yaml
import org.yaml.snakeyaml.constructor.Constructor
import org.yaml.snakeyaml.error.MarkedYAMLException
import org.yaml.snakeyaml.representer.Representer
import java.io.File
import java.util.*
import java.util.regex.Pattern

/**
 * Implements directive utilitary methods.
 *
 * @author Paulo Roberto Massa Cereda
 * @version 4.0
 * @since 4.0
 */
object DirectiveUtils {
    // the application messages obtained from the
    // language controller
    private val messages = LanguageController

    // get the logger context from a factory
    private val logger = LoggerFactory.getLogger(DirectiveUtils::class.java)

    /**
     * This function filters the lines of a file to identify the potential
     * directives.
     *
     * @param lines The lines of the file.
     * @return A map containing the line number and the line's content.
     */
    private fun getPotentialDirectiveLines(lines: List<String>):
            Map<Int, String> {
        val header = Arara.config[AraraSpec.Execution.onlyHeader]
        val validLineRegex = Arara.config[AraraSpec.Execution.filePattern]
        val validLinePattern = Pattern.compile(validLineRegex)
        val validLineStartPattern = (validLineRegex + Arara.config[AraraSpec
                .Application.namePattern]).toPattern()
        val map = mutableMapOf<Int, String>()
        for ((i, text) in lines.withIndex()) {
            val validLineMatcher = validLineStartPattern.matcher(text)
            if (validLineMatcher.find()) {
                val line = text.substring(validLineMatcher.end())
                map[i + 1] = line

                logger.info(messages.getMessage(
                        Messages.LOG_INFO_POTENTIAL_PATTERN_FOUND,
                        i + 1, line.trim()))
            } else if (header && !checkLinePattern(validLinePattern, text)) {
                // if we should only look within the file's header and reached
                // a point where the line pattern does not match anymore, we
                // assume we have left the header and break
                break
            }
        }
        return map
    }

    /**
     * Extracts a list of directives from a list of strings.
     *
     * @param lines List of strings.
     * @return A list of directives.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    @Throws(AraraException::class)
    @Suppress("MagicNumber")
    fun extractDirectives(lines: List<String>): List<Directive> {
        val pairs = getPotentialDirectiveLines(lines)
                .takeIf { it.isNotEmpty() }
                ?: throw AraraException(messages.getMessage(
                        Messages.ERROR_VALIDATE_NO_DIRECTIVES_FOUND))

        val assemblers = mutableListOf<DirectiveAssembler>()
        var assembler = DirectiveAssembler()
        val linebreakPattern = Arara.config[AraraSpec.Directive
                .linebreakPattern].toPattern()
        for ((lineno, content) in pairs) {
            val linebreakMatcher = linebreakPattern.matcher(content)
            if (linebreakMatcher.find()) {
                if (!assembler.isAppendAllowed) {
                    throw AraraException(
                            messages.getMessage(
                                    Messages.ERROR_VALIDATE_ORPHAN_LINEBREAK,
                                    lineno
                            )
                    )
                } else {
                    assembler.addLineNumber(lineno)
                    assembler.appendLine(linebreakMatcher.group(1))
                }
            } else {
                if (assembler.isAppendAllowed) {
                    assemblers.add(assembler)
                }
                assembler = DirectiveAssembler()
                assembler.addLineNumber(lineno)
                assembler.appendLine(content)
            }
        }
        if (assembler.isAppendAllowed) {
            assemblers.add(assembler)
        }

        return assemblers.map { generateDirective(it) }
    }

    /**
     * Generates a directive from a directive assembler.
     *
     * @param assembler The directive assembler.
     * @return The corresponding directive.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    @Throws(AraraException::class)
    @Suppress("MagicNumber")
    fun generateDirective(assembler: DirectiveAssembler): Directive {
        val regex = Arara.config[AraraSpec.Directive.directivePattern]
        val pattern = Pattern.compile(regex)
        val matcher = pattern.matcher(assembler.getText())
        if (matcher.find()) {
            val directive = Directive(
                    identifier = matcher.group(1)!!,
                    parameters = getParameters(matcher.group(3),
                            assembler.getLineNumbers()),
                    conditional = Conditional(
                            type = getType(matcher.group(5)),
                            condition = matcher.group(6) ?: ""
                    ),
                    lineNumbers = assembler.getLineNumbers()
            )

            logger.info(messages.getMessage(
                    Messages.LOG_INFO_POTENTIAL_DIRECTIVE_FOUND, directive))

            return directive
        } else {
            throw AraraException(
                    messages.getMessage(
                            Messages.ERROR_VALIDATE_INVALID_DIRECTIVE_FORMAT,
                            "(" + assembler.getLineNumbers()
                                    .joinToString(", ") + ")"
                    )
            )
        }
    }

    /**
     * Gets the conditional type based on the input string.
     *
     * @param text The input string.
     * @return The conditional type.
     */
    private fun getType(text: String?): Conditional.ConditionalType {
        return when (text) {
            null -> Conditional.ConditionalType.NONE
            "if" -> Conditional.ConditionalType.IF
            "while" -> Conditional.ConditionalType.WHILE
            "until" -> Conditional.ConditionalType.UNTIL
            else -> Conditional.ConditionalType.UNLESS
        }
    }

    /**
     * Gets the parameters from the input string.
     *
     * @param text    The input string.
     * @param numbers The list of line numbers.
     * @return A map containing the directive parameters.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    @Throws(AraraException::class)
    // TODO: loadAs might throw another (undocumented) exception
    private fun getParameters(text: String?,
                              numbers: List<Int>): Map<String, Any> {
        if (text == null)
            return mapOf()

        return Yaml(Constructor(),
                Representer(),
                DumperOptions(),
                DirectiveResolver()).runCatching {
            loadAs(text, HashMap::class.java).mapKeys { it.toString() }
        }.getOrElse {
            throw if (it is MarkedYAMLException)
                AraraException(messages.getMessage(
                        Messages.ERROR_VALIDATE_YAML_EXCEPTION,
                        "(" + numbers.joinToString(", ") + ")"),
                        it)
            else
                it
        }
    }

    /**
     * Replicate a directive for given files.
     *
     * @param holder The list of files.
     * @param parameters The parameters for the directive.
     * @param directive The directive to clone.
     * @return List of cloned directives.
     * @throws AraraException If there is an error validating the [holder]
     *   object.
     */
    @Throws(AraraException::class)
    private fun replicateDirective(holder: Any, parameters: Map<String, Any>,
                                   directive: Directive): List<Directive> {
        return if (holder is List<*>) {
            // we received a file list, so we map that list to files
            holder.filterIsInstance<Any>()
                    .asSequence()
                    .map { File(it.toString()) }
                    .map(CommonUtils::getCanonicalFile)
                    // and because we want directives, we replicate our
                    // directive to be applied to that file
                    .map { reference ->
                        directive.copy(parameters = parameters
                                .plus("reference" to reference))
                    }
                    .toList()
                    // we take the result if and only if we have at least one
                    // file and we did not filter out any invalid argument
                    .takeIf { it.isNotEmpty() && holder.size == it.size }
            // TODO: check exception according to condition
                    ?: throw AraraException(
                            messages.getMessage(
                                    Messages.ERROR_VALIDATE_EMPTY_FILES_LIST,
                                    "(" + directive.lineNumbers
                                            .joinToString(", ") + ")"
                            )
                    )
        } else {
            throw AraraException(
                    messages.getMessage(
                            Messages.ERROR_VALIDATE_FILES_IS_NOT_A_LIST,
                            "(" + directive.lineNumbers.joinToString(", ") + ")"
                    )
            )
        }
    }

    /**
     * Validates the list of directives, returning a new list.
     *
     * @param directives The list of directives.
     * @return A new list of directives.
     * @throws AraraException Something wrong happened, to be caught in the
     * higher levels.
     */
    @Throws(AraraException::class)
    fun process(directives: List<Directive>): List<Directive> {
        val result = mutableListOf<Directive>()
        directives.forEach { directive ->
            val parameters = directive.parameters

            if (parameters.containsKey("reference"))
                throw AraraException(messages.getMessage(
                        Messages.ERROR_VALIDATE_REFERENCE_IS_RESERVED,
                        "(" + directive.lineNumbers.joinToString(", ") + ")"))

            if (parameters.containsKey("files")) {
                result.addAll(replicateDirective(parameters.getValue("files"),
                        parameters.minus("files"), directive))
            } else {
                result.add(directive.copy(parameters = parameters
                        .plus("reference" to
                                Arara.config[AraraSpec.Execution.reference])))
            }
        }

        logger.info(messages.getMessage(
                Messages.LOG_INFO_VALIDATED_DIRECTIVES))
        logger.info(DisplayUtils.displayOutputSeparator(
                messages.getMessage(Messages.LOG_INFO_DIRECTIVES_BLOCK)))
        result.forEach { logger.info(it.toString()) }
        logger.info(DisplayUtils.displaySeparator())

        return result
    }

    /**
     * Checks if the provided line contains the corresponding pattern, based on
     * the file type, or an empty line.
     *
     * @param pattern Pattern to be matched, based on the file type.
     * @param line    Provided line.
     * @return Logical value indicating if the provided line contains the
     * corresponding pattern, based on the file type, or an empty line.
     */
    private fun checkLinePattern(pattern: Pattern, line: String): Boolean {
        return line.isBlank() || pattern.matcher(line).find()
    }
}
